### PR TITLE
[openblas] update to 0.3.15 version and add dynamic arch feature

### PR DESCRIPTION
**Pull request by xandox**
Add [dynamic arch](https://github.com/xianyi/OpenBLAS/#support-for-multiple-targets-in-a-single-library) feature to OpenBLAS ports

- What does your PR fix? Fixes # Add new feature

- Which triplets are supported/not supported? Have you updated the CI baseline? Same

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)? yes

### CODE REVIEW

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
@xandox 

Have you tested the feature `dynamic-arch`?

**Comment by xandox** [comment]
@NancyLi1013 before PR I build with dynamic-arch feature for x64-linux, today I try to build for arm64-android and version 0.3.10 would not build. I also update version.

**Comment by xandox** [comment]
Oh, how to debug failed tests?

**Comment by NancyLi1013** [comment]
For the failure on osx, there is no error log generated. I have rerun it and let's see the new test result.

For the feature test, you can test x86-windows and x64-windows.

File: ports/openblas/CONTROL [hunk 1]
```diff
@@ -1,5 +1,9 @@
 Source: openblas
 Version: 0.3.10
+Port-Version: 1
```
Comment (NancyLi1013) [inline ports/openblas/CONTROL:3]:
```suggestion
Version: 0.3.13
```

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
The failures are like this:

```
/Users/vagrant/Data/buildtrees/openblas/src/7079f26fa1-32aa3694bc.clean/kernel/x86_64/srot_microk_haswell-2.c:41:14: error: always_inline function '_mm256_fmadd_ps' requires target feature 'fma', but would be inlined into function 'srot_kernel' that is compiled without support for 'fma'
        t0 = _mm256_fmadd_ps(c_256, x0, t0);
             ^
/Users/vagrant/Data/buildtrees/openblas/src/7079f26fa1-32aa3694bc.clean/kernel/x86_64/srot_microk_haswell-2.c:42:14: error: always_inline function '_mm256_fmadd_ps' requires target feature 'fma', but would be inlined into function 'srot_kernel' that is compiled without support for 'fma'
        t1 = _mm256_fmadd_ps(c_256, x1, t1);
             ^
/Users/vagrant/Data/buildtrees/openblas/src/7079f26fa1-32aa3694bc.clean/kernel/x86_64/srot_microk_haswell-2.c:43:14: error: always_inline function '_mm256_fmadd_ps' requires target feature 'fma', but would be inlined into function 'srot_kernel' that is compiled without support for 'fma'
        t2 = _mm256_fmadd_ps(c_256, x2, t2);
```

**Comment by NancyLi1013** [comment]
The same error on x64-osx

**Comment by NancyLi1013** [comment]
Related PR #14760

**Comment by xandox** [comment]
Problem here is: `openblas` detects that processor (on check host) does not have `FMA` instructions set. And as it does not have this instructions `-mfma` flag does not append to compilator arguments. But `openblas` still use intrinsics from `FMA` set if processor is `HASWELL` and that is why error is produced.  
I created issue https://github.com/xianyi/OpenBLAS/issues/3128 for `openblas`.

**Comment by xandox** [comment]
Accidentally closed it. Missed the buttons

**Comment by xandox** [comment]
There a comment from @martin-frbg from openblas issue: 
> I do find it a bit strange that the cpu would report AVX/AVX2 capability but not FMA3 though - is this a VM config problem ?

Is it any chance that it can be fixed by configuring VM?

**Comment by NancyLi1013** [comment]
> There a comment from @martin-frbg from openblas issue:
> 
> > I do find it a bit strange that the cpu would report AVX/AVX2 capability but not FMA3 though - is this a VM config problem ?
> 
> Is it any chance that it can be fixed by configuring VM?

For fma problem, you can try to add an option for osx platform, it's unnecessary to configure VM.

` add_compile_options(-mfma)`

**Comment by xandox** [comment]
This is issue not about missed compiler flag or mac os itself. `openblas` perfectly compiles on my local mac. openblas crew already fix this problem but it will be in new release only. However, comment from @martin-frbg still applyable to VM.

**Comment by NancyLi1013** [comment]
@xandox 

Thanks for your above information. Is it possible to update `openblas ` to the latest commit or make a patch to fix the problem?

For the configuration on our CI machine, we will also check and confirm this later.

**Comment by xandox** [comment]
> Is it possible to update openblas to the latest commit or make a patch to fix the problem?

Yes. It looks like 0.3.14 is about to be released soon. I update this PR after this will happen.

**Comment by NancyLi1013** [comment]
Hi @xandox 

v0.3.14 has been released now. https://github.com/xianyi/OpenBLAS/releases/tag/v0.3.14

Could you please update to 0.3.14 in this PR?

**Comment by xandox** [comment]
(╯°□°)╯︵ ┻━┻

**Comment by NancyLi1013** [comment]
Seems the new release doesn't fix the problem. 

The same error as before:
```
/Users/vagrant/Data/buildtrees/openblas/src/2e36dd2a71-ba2c41d242.clean/kernel/x86_64/drot_microk_haswell-2.c:41:14: error: always_inline function '_mm256_fmadd_pd' requires target feature 'fma', but would be inlined into function 'drot_kernel' that is compiled without support for 'fma'
        t0 = _mm256_fmadd_pd(c_256, x0, t0);
```

File: versions/o-/openblas.json [hunk 2]
```diff
@@ -1,5 +1,20 @@
...
+    {
+      "git-tree": "f9dd03688cf5141bedd414dc492c4754ac640e21",
+      "version-string": "0.3.13",
+      "port-version": 0
+    },
+    {
+      "git-tree": "b8be2a38eb32ecdd979fbefa78c7933436924c0d",
+      "version-string": "0.3.10",
+      "port-version": 1
+    },
```
Comment (NancyLi1013) [inline versions/o-/openblas.json:17]:
```suggestion
```
There should be only one change for every PR.

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by martin-frbg** [comment]
> Seems the new release doesn't fix the problem.
> 
> The same error as before:
> 
> ```
> /Users/vagrant/Data/buildtrees/openblas/src/2e36dd2a71-ba2c41d242.clean/kernel/x86_64/drot_microk_haswell-2.c:41:14: error: always_inline function '_mm256_fmadd_pd' requires target feature 'fma', but would be inlined into function 'drot_kernel' that is compiled without support for 'fma'
>         t0 = _mm256_fmadd_pd(c_256, x0, t0);


> Seems the new release doesn't fix the problem.
> 
> The same error as before:
> 
> ```
> /Users/vagrant/Data/buildtrees/openblas/src/2e36dd2a71-ba2c41d242.clean/kernel/x86_64/drot_microk_haswell-2.c:41:14: error: always_inline function '_mm256_fmadd_pd' requires target feature 'fma', but would be inlined into function 'drot_kernel' that is compiled without support for 'fma'
>         t0 = _mm256_fmadd_pd(c_256, x0, t0);
> ```

can you post the config.h generated during the failed build please ?

**Comment by xandox** [comment]
unfortunately `config.h` not in artifacts. @NancyLi1013 can you help with this?

**Comment by martin-frbg** [comment]
nevermind - I realized that I had only fixed the srot microkernel while the same mistake exists in drot as well.

**Comment by gnaggnoyil** [comment]
It seems that version bump and dynamic arch are two separate features that shouldn't impact on each other. As a user, we are waiting to use v0.3.14 in vcpkg thus we think it would be great if there's a separate PR that bumps version and make this PR rebase on the version-bumped one so that this PR would focus on dynamic arch issues.

**Comment by xandox** [comment]
Unfortunately `dynamic-arch` feature is not issue here. It can't be compiled in core mode

**Comment by gnaggnoyil** [comment]
@xandox So can we ignore this and bump version first?

**Comment by martin-frbg** [comment]
can you wait a bit (until the weekend I hope) and go straight to 0.3.15 then ?

**Comment by xandox** [comment]
>So can we ignore this and bump version first?
No, problem in new versions. Probably if set dynamic-arch as default feature - it can helps, but I do not know is it good idea?

**Comment by xandox** [comment]
@martin-frbg it will be grate!

**Comment by gnaggnoyil** [comment]
@xandox @martin-frbg Sure. I'll wait for the 0.3.15 version in vcpkg.

**Comment by martin-frbg** [comment]
0.3.15 released now

**Comment by xandox** [comment]
@martin-frbg thank you!

**Comment by xandox** [comment]
up

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
LGTM now, thanks for your patience and contribution @xandox. Also thanks everyone here.

File: ports/openblas/portfile.cmake [hunk 3]
```diff
@@ -27,6 +26,18 @@ vcpkg_add_to_path("${SED_EXE_PATH}")
...
 set(COMMON_OPTIONS -DBUILD_WITHOUT_LAPACK=ON)
 
+vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+    FEATURES
+    	"dynamic-arch"      DYNAMIC_ARCH
+)
+
+if(VCPKG_TARGET_IS_OSX)
+    if("dynamic-arch" IN_LIST FEATURES)
+        set(VCPKG_LIBRARY_LINKAGE dynamic)
```
Comment (dan-shaw) [inline ports/openblas/portfile.cmake:36]: We generally should not change library linkage if a feature is enabled. If a feature does not support a linkage type, we should output `message(FATAL_ERROR)`. Does the `dynamic-arch` feature work for other platforms? We should also output an error if this feature is enabled on platforms that are not supported.

Comment (xandox) [reply ports/openblas/portfile.cmake:36]: There actually a issue not with `openblas` directly but in ninja.
https://gitlab.kitware.com/cmake/cmake/-/issues/16731
rsp-file contains 7001 lines of other libraries. 

Can you suggest some other ways to resolve it without `message(FATAL_ERROR)`?

Comment (ras0219) [reply ports/openblas/portfile.cmake:36]:
```
vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)
```

plus a message about why this is happening.

**Comment by NancyLi1013** [comment]
@xandox 
Could you please resolve the conflicts in this PR?

**Comment by xandox** [comment]
> @xandox
> Could you please resolve the conflicts in this PR?

done, once again

**Comment by NancyLi1013** [comment]
LGTM now, thanks for your PR. @xandox

**Review by vicroms — APPROVED** [review]
This LGTM once merge conflicts have been solved

**Pull request merged**